cpp/src/tests/community/ecg_test.cu

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * NVIDIA CORPORATION and its licensors retain all intellectual property
+ * and proprietary rights in and to this software, related documentation
+ * and any modifications thereto.  Any use, reproduction, disclosure or
+ * distribution of this software and related documentation without an express
+ * license agreement from NVIDIA CORPORATION is strictly prohibited.
+ *
+ */
+#include <gtest/gtest.h>
+#include <nvgraph/nvgraph.h>
+#include <cugraph.h>
+#include <algorithm>
+#include "test_utils.h"
+
+#include <rmm_utils.h>
+
+TEST(nvgraph_louvain, success)
+{
+  cugraph::Graph G;
+
+  std::vector<int> off_h = {0, 16, 25, 35, 41, 44, 48, 52, 56, 61, 63, 66, 67, 69, 74, 76, 78, 80, 82, 84, 87, 89, 91, 93, 98, 101, 104, 106, 110, 113, 117, 121, 127, 139, 156};
+  std::vector<int> ind_h = {1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 17, 19, 21, 31, 0, 2, 3, 7, 13, 17, 19, 21, 30, 0, 1, 3, 7, 8, 9, 13, 27, 28, 32, 0, 1, 2, 7, 12, 13, 0, 6, 10, 0,
+      6, 10, 16, 0, 4, 5, 16, 0, 1, 2, 3, 0, 2, 30, 32, 33, 2, 33, 0, 4, 5, 0, 0, 3, 0, 1, 2, 3, 33, 32, 33, 32, 33, 5, 6, 0, 1, 32, 33, 0, 1, 33, 32, 33, 0, 1, 32, 33, 25, 27, 29, 32, 33,
+      25, 27, 31, 23, 24, 31, 29, 33, 2, 23, 24, 33, 2, 31, 33, 23, 26, 32, 33, 1, 8, 32, 33, 0, 24, 25, 28, 32, 33, 2, 8, 14, 15, 18, 20, 22, 23, 29, 30, 31, 33, 8, 9, 13, 14, 15,
+      18, 19, 20, 22, 23, 26, 27, 28, 29, 30, 31, 32};
+  std::vector<float> w_h = {1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
+      1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
+      1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
+      1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
+      1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0};
+
+  gdf_column col_off, col_ind, col_w;
+
+
+  create_gdf_column(off_h,&col_off);
+  create_gdf_column(ind_h,&col_ind);
+  create_gdf_column(w_h  ,&col_w);
+
+  cugraph::adj_list_view(&G, &col_off, &col_ind, &col_w);
+
+  int no_vertex = off_h.size()-1;
+  int* best_cluster_vec = NULL;
+
+  cudaStream_t stream{nullptr};
+  ALLOC_TRY((void**)&best_cluster_vec, sizeof(int) * no_vertex, stream);
+  gdf_column result_col;
+  gdf_column_view(&result_col, best_cluster_vec, nullptr, no_vertex, GDF_INT32);
+
+  ASSERT_NO_THROW(cugraph::ecg(&G, .05, 16, &result_col));
+
```

Comment:
I am not sure how to compare it to anything in the C test... The python tests do such comparisons so I don't think it's necessary to do them here. We could remove the test entirely if you think it's better to not have it.

Reply:
The C++ test needs to verify that what is returned by the C++ API is valid at this level. 

A valid clustering would have V entries, all positive, all smaller than V, etc.

It would be reasonable to analyze the modularity score of a random permutation of size V and expect it to be lower than the score resulting of ECG clustering. I think it should be a dozen lines of code and can save debugging time later on.

A possible alternative is to compare to the know modularity score, like in the C++ test of Louvain.

Reply:
Updated the c++ test to compute and check modularity score and check that the returned partition ids are dense.